### PR TITLE
unravel_index

### PR DESCRIPTION
**Pull request by Hakuyume**
This PR implements `cupy.unravel_index`.

### CODE REVIEW

**Comment by unnonouno** [comment]
jenkins test this please

File: cupy/indexing/generate.py [hunk 1]
```diff
@@ -252,7 +252,57 @@ def ix_(*args):
 # TODO(okuta): Implement ravel_multi_index
 
 
-# TODO(okuta): Implement unravel_index
+def unravel_index(indices, dims, order='C'):
+    """Converts a flat index or array of flat indices into a tuple of coordinate arrays.
+
+    Args:
```
Comment (niboshi) [inline cupy/indexing/generate.py]: Please describe accepted types (or values) of arguments.

File: cupy/indexing/generate.py [hunk 2]
```diff
@@ -252,7 +252,57 @@ def ix_(*args):
...
+
+    .. seealso:: :func:`numpy.unravel_index`
+
+    """
+    if order == 'C':
+        dims = reversed(dims)
+    elif order == 'F':
+        pass
+    else:
+        raise TypeError("order not understood")
```
Comment (niboshi) [inline cupy/indexing/generate.py]: Please use single quotes. (Applies for other string literals as well).

File: cupy/indexing/generate.py [hunk 3]
```diff
@@ -252,7 +252,57 @@ def ix_(*args):
...
+        raise TypeError(
+            "Iterator operand 0 dtype could not be cast "
+            "from dtype('{}') to dtype('{}') "
+            "according to the rule 'same_kind'".format(
+                indices.dtype, cupy.int64().dtype))
+
+    if (indices < 0).any():
+        raise ValueError("invalid entry in index array")
+
+    unraveled_coords = list()
```
Comment (niboshi) [inline cupy/indexing/generate.py]: Use `[]`.

File: tests/cupy_tests/indexing_tests/test_generate.py [hunk 4]
```diff
@@ -149,3 +149,41 @@ def test_AxisConcatenator_init1(self):
...
+@testing.gpu
+class TestUnravelIndex(unittest.TestCase):
+
+    _multiprocess_can_split_ = True
+
+    @testing.for_int_dtypes()
+    @testing.numpy_cupy_array_list_equal()
+    def test_C(self, xp, dtype):
+        a = testing.shaped_arange((4, 3, 2), xp, dtype)
+        a = xp.minimum(a, 6 * 4 - 1)
```
Comment (niboshi) [inline tests/cupy_tests/indexing_tests/test_generate.py]: Is this `minimum` needed?

File: tests/cupy_tests/indexing_tests/test_generate.py [hunk 5]
```diff
@@ -149,3 +149,41 @@ def test_AxisConcatenator_init1(self):
...
+        return xp.unravel_index(a, (6, 4))
+
+    @testing.for_int_dtypes()
+    @testing.numpy_cupy_array_list_equal()
+    def test_F(self, xp, dtype):
+        a = testing.shaped_arange((4, 3, 2), xp, dtype)
+        a = xp.minimum(a, 6 * 4 - 1)
```
Comment (niboshi) [inline tests/cupy_tests/indexing_tests/test_generate.py]: Ditto.

File: tests/cupy_tests/indexing_tests/test_generate.py (hunk 4 above)
Comment (Hakuyume) [reply tests/cupy_tests/indexing_tests/test_generate.py]: Without this `minimum`, `a` contains `24`. This index is invalid.

File: cupy/indexing/generate.py (hunk 1 above)
Comment (Hakuyume) [reply cupy/indexing/generate.py]: I followed the style of other functions' docs. For example, the docs of `ix_` does not have type description. I thought that was cupy style.

File: cupy/indexing/generate.py (hunk 2 above)
Comment (Hakuyume) [reply cupy/indexing/generate.py]: I also prefer single quotes. However, the docs of other functions use double quotes. So I followed them.

File: cupy/indexing/generate.py (hunk 3 above)
Comment (Hakuyume) [reply cupy/indexing/generate.py]: OK. Is this coding rule documented in Chainer/CuPy?

File: cupy/indexing/generate.py (hunk 1 above)
Comment (niboshi) [reply cupy/indexing/generate.py]: Even though it's missing in some functions, we should document it as much as possible.

File: cupy/indexing/generate.py (hunk 2 above)
Comment (niboshi) [reply cupy/indexing/generate.py]: We use triple double quotes in docstrings and use single quotes for normal string literals.

Comment (Hakuyume) [reply cupy/indexing/generate.py]: https://github.com/cupy/cupy/blob/master/cupy/indexing/generate.py#L242

File: cupy/indexing/generate.py (hunk 1 above)
Comment (Hakuyume) [reply cupy/indexing/generate.py]: OK.

File: cupy/indexing/generate.py (hunk 2 above)
Comment (niboshi) [reply cupy/indexing/generate.py]: That's against the style of other part of code and something we should fix.

Comment (Hakuyume) [reply cupy/indexing/generate.py]: OK. I think all rules which you pointed are not documented in https://docs.chainer.org/en/stable/contribution.html

File: tests/cupy_tests/indexing_tests/test_generate.py (hunk 4 above)
Comment (niboshi) [reply tests/cupy_tests/indexing_tests/test_generate.py]: OK.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**